walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,188 @@
 
 package com.google.sps;
 
+import com.google.sps.TimeRange;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  public static final int WHOLE_DAY = 24 * 60;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> meetings = new ArrayList<>();
+    Collection<String> requestedAttendees = request.getAttendees();
+    Collection<String> requestedOptionalAttendees = 
+      request.getOptionalAttendees();
+    long durationOfRequest = request.getDuration();
+
+    // Case 0: if duration too long (longer than a day), return no options
+    if (durationOfRequest > WHOLE_DAY) {
+      return meetings;
+    }
+
+    /** Case 1: GENERAL CASE >> remove conflicting times(from @events) 
+      * from possible meetings
+      */
+    meetings.add(TimeRange.fromStartDuration(0, WHOLE_DAY));
```

Comment:
This is meant to start a conversation instead of a request to change. I wanted to talk through your design a bit.

Reply:
I feel that the advantage is that my current implementation keeps my code flexible by making use of TimeRange class and it's preexisting attributes. If I use BitSet, I would have to convert timeranges into bitsets, and then eventually convert them back into TimeRange(s) if I want to manipulate them during the meeting-making process. 
However, thinking about object oriented programming and the longevity of the design, I was thinking it might be better to keep the process simple by using only TimeRange and not wrapping the data around anything else. That way, if I needed to change or add to the process of making new proposed meetings in the future, then at any point in the process my data will still have access to TimeRange properties such as contains(), overlaps(), and the comparator ORDER_BY_END / ORDER_BY_START.

Reply:
Let me know what you think about this idea!

Reply:
The meetings ArrayList represents the times when a new event can be scheduled.  It is currently represented by a pair of (start, end) times which for the most part is a compact representation.  By using this representation, you have to complicated logic for merging the current meetings array list with the next meeting in `makeNewProposedMeetings` which has 4 different cases with if/else logic.  

What I was proposing (and might be a silly idea) is to instead represent the day at a set of bits which represent if that minute is free to schedule a meeting.  So to fill this 

```
start_minute = ConvertTimeToMinutesFromMidnigth(time.getStartTime());
end_minute = ConvertTimeToMinutesFromMidnigth(time.getStartTime());
meetings[start_minute:end_minute] = false;
```

Now to find a time you can schedule the meeting all you would have to do is do one loop through the meetings bit set to see if there was a block large enough.  This seems simpler than the if/else logic, but it makes the tradeoff that it has more space complexity. 

You could also think about using this representation that minimizes the number of conflicts.  So instead of storing a boolean you could store the number of conflicts. Then you could try to find the minimum interval sum.

Reply:
I see, so what you're saying is with this bitset approach it's kind of more continuous in the sense that the data structure isn't split up into separate objects that contain each possible timerange(?) 

I liked the if/else logic because it helped me to visualize the process and to better think about how to filter for a possible timerange.

But I see how the approach you suggested also makes sense. In that case, should I go ahead and implement your suggestion?

Reply:
One other thing that we can consider too is that the bitset approach might be slower than my current implementation. From what I understand, the bitset would be 1440 elements in length, and I would be looping through all of that for each event, whereas with my implementation I am just looping through my meetings array which starts off with a best case of 1 itteration(since the initial length is 1), and as the proposed meeting times are split off into smaller interval, the meetings array increase by a worst case of +2 elements for each call of makeNewProposedMeetings(). However, I'm also not sure what we want our main focus to be, whether it is space complexity or time complexity. 

Therefore, please get back to me with your thoughts on this.

Reply:
No need to change your implementation.  I wanted to use this opportunity to talk through an alternative approach.

Reply:
Oh, ok, sounds good. I also went ahead and spent some time to code the alternative approach this morning. The code is there just for future reference in case I want to do something similar in the future. I didn't implement the code, but it is viewable as commented out code in a dummy file called AlternateFindMeetingQueryDesign.java